utils/metered_cache.go

```diff
@@ -0,0 +1,114 @@
+// (c) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"fmt"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/ava-labs/subnet-evm/metrics"
+)
+
+// MeteredCache wraps *fastcache.Cache and periodically pulls stats from it.
+type MeteredCache struct {
+	*fastcache.Cache
+	namespace string
+
+	// stats to be surfaced
+	entriesCount metrics.Gauge
+	bytesSize    metrics.Gauge
+	collisions   metrics.Gauge
+	gets         metrics.Gauge
+	sets         metrics.Gauge
+	misses       metrics.Gauge
+
+	// count all operations to decide when to update stats
+	ops             uint64
+	updateFrequency uint64
+}
+
+// NewMeteredCache returns a new MeteredCache that will update stats to the
+// provided namespace once per each [updateFrequency] operations.
+// Note: if [updateFrequency] is passed as 0, it will be treated as 1.
+func NewMeteredCache(size int, journal string, namespace string, updateFrequency uint64) *MeteredCache {
+	var cache *fastcache.Cache
+	if journal == "" {
+		cache = fastcache.New(size)
+	} else {
+		cache = fastcache.LoadFromFileOrNew(journal, size)
+	}
+	if updateFrequency == 0 {
+		updateFrequency = 1 // avoid division by zero
+	}
+	mc := &MeteredCache{
+		Cache:           cache,
+		namespace:       namespace,
+		updateFrequency: updateFrequency,
+	}
+	if namespace != "" {
+		// only register stats if a namespace is provided.
+		mc.entriesCount = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/entriesCount", namespace), nil)
+		mc.bytesSize = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/bytesSize", namespace), nil)
+		mc.collisions = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/collisions", namespace), nil)
+		mc.gets = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/gets", namespace), nil)
+		mc.sets = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/sets", namespace), nil)
+		mc.misses = metrics.GetOrRegisterGauge(fmt.Sprintf("%s/misses", namespace), nil)
+	}
+	return mc
+}
+
+// updateStats updates metrics from fastcache
+func (mc *MeteredCache) updateStatsIfNeeded() {
+	if mc.namespace == "" {
+		return
+	}
+	mc.ops++
+	if mc.ops%mc.updateFrequency != 0 {
+		return
+	}
+
+	s := fastcache.Stats{}
+	mc.UpdateStats(&s)
```

Comment:
Do we want to store a metric for how long this takes to run?

Reply:
made a change. or did you mean a cumulative metric to measure the total time spent updating stats?

Reply:
I meant the latter (want to make sure this doesn't take too long)